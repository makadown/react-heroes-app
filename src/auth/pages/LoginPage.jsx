import React, { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {
  
  // retrieving the login method from AuthProvider.jsx via useContext
  const {login} = useContext(AuthContext);

  const navigate = useNavigate();

  const onLogin = () => {

    login('Ampy Magny');

    navigate('/marvel', {
      replace: true
    });
  }

  return (
      <>
      <div className="container mt-5">
        <h1>Login</h1>
        <hr />
        <button
            className='btn btn-primary'
            onClick={onLogin}
            >
              Login
        </button>

      </div>
      </>
  )
}
