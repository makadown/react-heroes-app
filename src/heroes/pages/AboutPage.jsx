import React from 'react'

export const AboutPage = () => {
  return (
    <section className="header03 cid-rVmJmn8KSM mt-1 p-5" id="header03-2k">   

    <div className="mbr-overlay" style={{opacity: 0.5, backgroundColor: '#000000'}}></div>

    <div className="container align-left">
        <div className="row justify-content-left">
            <div className="mbr-white col-md-12 col-lg-6">
                <h1 className="mbr-section-title mbr-bold pb-3 mbr-fonts-style display-6">ABOUT</h1>
                <p className="mbr-section-subtitle mbr-light mbr-fonts-style">
                  Proyecto hecho en React 18 (powered by Vite). Conceptos practicados:</p>
                     <ul className='mt-2'>
                         <li>React Router</li>
                         <li>Navigate</li>
                         <li>useMemo</li>
                         <li>animaciones</li>
                         <li>protecci√≥n de rutas</li>
                         <li>login</li>
                         <li>Context</li>
                         <li>Reducer</li>
                     </ul>
                  Parte del 
                        <a href="https://www.udemy.com/course/react-cero-experto" 
                          className='ms-1' 
                          target='_blank'> 
                        curso del Maestro Fernando Herrera 
                        </a>.
                  
                  <p className="mbr-section-subtitle mbr-light mbr-fonts-style">
                  Alumno: Mario Arturo Serrano Flores.
                  </p>
                
                <div className="mbr-section-btn">
                  <a className="btn btn-info" href="https://github.com/makadown" target='_blank'>
                  
                  <svg xmlns="http://www.w3.org/2000/svg" 
                      width="20" 
                      height="20" 
                      fill="currentColor" 
                      className="bi bi-github me-2 mb-1" 
                      viewBox="0 0 16 16">
                    <path d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27s1.36.09 2 .27c1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.01 8.01 0 0 0 16 8c0-4.42-3.58-8-8-8"/>
                  </svg>    
                  My Repos</a></div>
            </div>
        </div>
    </div>
</section>
  )
}
