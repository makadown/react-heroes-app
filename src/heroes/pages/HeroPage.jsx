import React, { useEffect, useMemo, useState } from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroById } from '../helpers';

export const HeroPage = () => {

  const {id} = useParams();
  const navigate = useNavigate();

  const hero = useMemo(() => getHeroById(id), [id]);

  // Showing the hero info, also show a button to go back to home ('/')
  return (
    <>
     { 
      hero &&
      <div className='row mt-2 ml-2 mb-3'>
        <div className='col-4'>
          <img src={`/assets/heroes/${id}.jpg`} alt={hero.superhero} className='img-thumbnail animate__animated animate__fadeInLeft' />
        </div>
        <div className='col-8 ml-5 text-white bg-dark rounded'>
          <h3 className='mt-5 ms-2'>{hero.superhero}</h3>
          <ul className='list-group list-group-flush text-white bg-dark rounded-lg'>
            <li className='list-group-item text-white bg-dark ms-2 me-3'><b>Alter ego: </b>{hero.alter_ego}</li>
            <li className='list-group-item text-white bg-dark ms-2 me-3'><b>Publisher: </b>{hero.publisher}</li>
            <li className='list-group-item text-white bg-dark ms-2 me-3'><b>First appearance: </b>{hero.first_appearance}</li>
          </ul>
          <h5 className='mt-3 ms-2'>Characters</h5>
          <p className='ms-2 me-3'>{hero.characters}</p>
          <button className='btn btn-outline-info mb-3' onClick={() => navigate(-1)}>Back</button>          
        </div>
      </div>
     }
     {
      !hero &&
      <Navigate to="/"/>
     }
    </>
  )
}
