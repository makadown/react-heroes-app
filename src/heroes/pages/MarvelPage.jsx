import React from 'react';
import { HeroList } from '../components/HeroList';

export const MarvelPage = () => {
  return (
    <div>
      <h1>Marvel Page</h1>
      <hr />
      <HeroList publisher={'Marvel Comics'} />      
    </div>
  )
}
