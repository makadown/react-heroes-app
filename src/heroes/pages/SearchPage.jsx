import React from 'react';
import { HeroCard } from '../components/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { getHeroesByName } from '../helpers'; 

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);
  const heroes = q === '' ? [] : getHeroesByName(q);

  const { searchText, onInputChange } = useForm({
    searchText: q
  });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    navigate(`?q=${searchText}`);
  }

  return (
    <>
      <h1>Search Page</h1>
      <hr />
      <div className='row'>
        <div className='col-5'>
          <h4>Find your hero</h4>
          <hr />
          <form onSubmit={ onSearchSubmit }>
            <input type="text"
              placeholder="type here your hero's name"
              className='form-control'
              name="searchText"
              autoComplete='off'
              value={searchText}
              onChange={onInputChange}
            />
            <button className='btn btn-outline-info mt-2'>Search</button>
          </form>
        </div>
        <div className='col-7'>
          <h4>Results</h4>
          <hr />
          {
           q === '' &&
            <div className='alert alert-info animate__animated animate__fadeIn'>Search a hero</div>
          }
          {
            heroes.length === 0 && q.length > 1 &&
                <div className='alert alert-danger animate__animated animate__fadeIn'> No hero { q } found </div>
          }
          {
            heroes.length > 0 && 
              heroes.map(hero => (
                <HeroCard key={hero.id} {...hero} />
              ))
          }
        </div>
      </div>
    </>
  )
}
