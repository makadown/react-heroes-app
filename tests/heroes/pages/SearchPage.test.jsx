
/**
 * This is the test for SearchPage.jsx in src/heroes/pages
 * 1. Should render the search form
 * 2. Should render the heroes list
 */

import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';


const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
   useNavigate: () => mockedUsedNavigate
}));

describe('SearchPage tests', () => {

    beforeEach(() => jest.clearAllMocks());

    test('Should render the search form', () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );
        expect(screen.getByText('Search a hero')).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    test('Should render Batman and the input with the queryString value', () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman'] }>
                <SearchPage />
            </MemoryRouter>
        );
        
        // checking input with queryString value
        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');
        // look for an image in the DOM to verify that there is at least 1 img with class "card-img"
        expect(screen.getAllByRole('img').length).toBeGreaterThan(0);

        const img = screen.getByRole('img');
        expect(img.src).toContain('assets/heroes/dc-batman.jpg');

        // check that HeroCard component is rendered
        const heroCard = screen.getByLabelText('hero-card');
        expect(heroCard).toBeTruthy();        
    });

    test('Should show an error if the hero is not found', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman123'] }>
                <SearchPage />
            </MemoryRouter>
        );

        expect(screen.getByText('No hero batman123 found')).toBeTruthy();
    });

   test('Should call the useNavigate hook', () => {
        render(
            <MemoryRouter initialEntries={['/search'] }>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { value: 'batman' } });
        const form = screen.getByRole('form');
        fireEvent.submit(form);
        expect(mockedUsedNavigate).toHaveBeenCalledWith('?q=batman');
    });

});

