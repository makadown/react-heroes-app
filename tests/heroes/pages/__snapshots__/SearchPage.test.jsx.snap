// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage tests Should render the search form 1`] = `
<div>
  <h1>
    Search Page
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Find your hero
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="type here your hero's name"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-info mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
